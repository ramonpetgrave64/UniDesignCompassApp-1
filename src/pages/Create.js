/**
 * @fileoverview The Create Page that allows logged in users to create their Compasses
 * @author Abraham Villaroman
 * @version 1.0.0
*/

import React from "react";
import { navigate } from 'gatsby';
import Layout from '../components/layout';
import { Button, Form } from 'react-bootstrap';
import _ from "lodash";
import {createNewCompass} from '../graphql_utils/utils';
import { connect } from 'react-redux';
import { updateUser } from '../state/actions'
import SpinningWheel from "../components/SpinningWheel"

/**
 * @class SubmitCompass
 * Form that handles submission of the design process
 * takes in the title of the desired compass and fires the createCompass handler
 */
export const SubmitCompass = (props) =>  (
  <Form.Group controlId="formBasicEmail">
    <Form.Label className="form-title">Title of your Design Compass</Form.Label>
    <div className="input-area">
      <Form.Control name="title" type="text" placeholder="Title" onChange={props.onChange} value={props.title} />
      <Button id="create-button" className="input-button" variant="primary" onClick={props.createCompass}> Submit Compass </Button>
    </div>
  </Form.Group>
)

/**
 * @class Choose Structure
 * Form that handles Structure of the design process
 * prompts the user to either pick the default structure or a custom structure
 */
export const ChooseStructure = (props) =>  (
    <Form.Group controlId="formBasicEmail">
      <Form.Label className="form-title">Choose a Compass Style:</Form.Label>
      <div className="input-area">
        <Button className="input-button" id="Custom"  variant="primary" onClick={e => {props.handleCompassType(false)}}> Custom </Button>
        <Button className="input-button" id="Default" variant="primary" onClick={e => {props.handleCompassType(true)}}> Default </Button>
      </div>
      
    </Form.Group>
)
const default_phases= [
  {title:"Define a Problem",
   description:"Phase for defining the problem"},
   {title:"Research",
   description:"researching our problem"},
   {title:"Brainstorm",
   description:"brain storming solutions"},
   {title:"Select",
   description:"selecting the right solutions"},
   {title:"Construct",
   description:" constructing the solution"},
   {title:"Evaluate",
   description:" evaluating the implmentation of the solution"},
   {title:"Communicate",
   description:"test solution with others"},
   {title:"Redesign",
   description:"redesigning thesolution"}
]

/**
 * @class PhaseNumber
 * Form that handles Structure of the design process
 * prompts the user to choose the number phases they want in their process
 */
export const PhaseNumber = (props) =>  (
  <Form.Group controlId="exampleForm.ControlSelect1">
    <Form.Label className="form-title">How many phases are in your design process</Form.Label>
    <div className="input-area">
      <Form.Control as="select" onChange={props.handlePhaseAmount}>
        <option value="0">-</option>
        <option value="2">2</option>
        <option value="3">3</option>
        <option value="4">4</option>
        <option value="5">5</option>
        <option value="6">6</option>
        <option value="7">7</option>
        <option value="8">8</option>
        <option value="9">9</option>
        <option value="10">10</option>
        <option value="11">11</option>
        <option value="12">12</option>
      </Form.Control>
    </div>
  </Form.Group>
)

/**
 * Form that handles creation of a phase in their design process
 * prompts the user to name their phase and it's description
 */
export class DescribePhase extends React.Component{
  state = {
    title: "",
    description: "",
  }

  onChange = (e) => {
    this.setState({
      [e.target.name] : e.target.value 
    },() => {
      this.props.createPhase(this.state.title, this.state.description,this.props.id + 1)
    })

  }

  render(){
    const {title,description} = this.state;
    return (
      <Form.Group >
        <Form.Label className="form-sub-title">{`Phase ${this.props.id + 1}`}</Form.Label>
        <Form.Control id="phaseTitle" name="title" type="text" placeholder="Title" onChange={this.onChange} value={title} />
        <Form.Control id="phaseDescription" name="description" as="textarea" placeholder="Description" onChange={this.onChange} value={description} rows="3" />
      </Form.Group>
    )
  }
} 

/**
 * The page that handles creation of a a design process
 * holds all the logic in changing and storing the form data, then submits the data to the server
 */
export class CreatePage extends React.Component {
  state = {
    compassTitle: "",
    canSubmit: false,
    defaultCompass: true,
    numberOfPhases: 0,
    phases: [],
    status: "chooseStructure",
    loading: false
  }

/**
 * event handler for the Process title
 * changes the value of this.state.compassTitle
 * @param {Object} event object that holds the value of the Process Title
*/
  handleCompassTitle = (e) => {
    this.setState({compassTitle: e.target.value})
  }

/**
 * event handler for the Process Structure
 * changes the value of this.state.defaultCompass
 * when the user submits the structure, switch to the submitCompass form
 * @param {bool} boolean that holds the structure type of our Process (true === default structure, false === custom structure)
*/
  handleCompassType = (isDefaultCompass) => {
    if (isDefaultCompass){
      this.setState({defaultCompass: isDefaultCompass, phases: default_phases,status: "submitCompass"})
    } else {
      this.setState({defaultCompass: isDefaultCompass, status: "numOfPhases"})
    }
  }

/**
 * event handler for the Phase amount 
 * changes the value of this.state.phaseAmount
 * when the user submits the number of phases, switch to the createPhases form
 * @param {Object} event object that holds the value of the Phase Amount
*/
  handlePhaseAmount = (e) => {
    this.setState({numberOfPhases: e.target.value,status:"createPhases"})
  }

/**
 * event handler that creates the phases
 * changes the value of this.state.phases with an array of phases with their title and descriptions
 * @param {Object} event object that holds the Phase objects
*/
  createPhase = (title,description,phaseNumber) => {
    let phases = this.state.phases
    phases.push({title,description,phaseNumber});
    phases = _.uniqBy(phases.reverse(), 'phaseNumber')
    this.setState({phases})
  }
  
/**
 * makes the api request to submit a new compass with: Process Title and the defined phases in the Process
 * if the server request is succesful, the user is redirected to the process screen
 * otherwise, there was an error
*/
  createCompass = () => {
    const {compassTitle, phases} = this.state;
    this.setState({ loading: true }, () => {
      createNewCompass(this.props.user,compassTitle,phases)
      .then((res) => {
          navigate(`/Process/${res.data.getProcess.id}`)
      })
      .catch((err) => {
        alert(err)
        this.forceUpdate();
      })
    })
  }
  /**
   * when the user is done with all the phases, change the form to the submitCompass Form
   */
  submitPhases = (e) => {
    this.setState({status: "submitCompass"})
  }

  /**
   * handles the displaying of forms for the user to operate with to create their design process
   */
  handleForms = () => {
    switch(this.state.status) { 
      case "chooseStructure":
        return <ChooseStructure handleCompassType={this.handleCompassType}/>   
      case "numOfPhases":
        return <PhaseNumber handlePhaseAmount={this.handlePhaseAmount}/>
      case "createPhases":
        const Phases = Array.from({length: this.state.numberOfPhases}, (_, id) => ( 
            <DescribePhase createPhase={this.createPhase} id={id} key={id}/> 
          )
        )
        return (
          <div>
            <Form.Label className="form-title">Describe each phase in your design process</Form.Label>
              {Phases}
            <Button onClick={this.submitPhases}>Submit Phases</Button>
          </div>
        )
      case "submitCompass":
        return <SubmitCompass 
                  createCompass={this.createCompass} 
                  title={this.state.compassTitle} 
                  onChange={this.handleCompassTitle}
                />
      default: 
        return <ChooseStructure handleCompassType={this.handleCompassType}/>   
    }
  }

  render(){
    return (
      <Layout>
          <h2> Create Your Compass </h2>
          <div className="form-container">
            {this.handleForms()}
            { this.state.loading && <SpinningWheel/>}
          </div>
      </Layout>
    )
  }
}
const mapStateToProps =({state})=>(
  {user: state.user}
)
const mapDispatchToProps = dispatch => ({
  updateUser: (user) => dispatch(updateUser(user))
})
export default connect(mapStateToProps,mapDispatchToProps)(CreatePage);
