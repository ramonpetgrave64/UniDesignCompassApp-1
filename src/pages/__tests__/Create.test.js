import React from "react"
import renderer from "react-test-renderer"
import {configure, mount, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {CreatePage, DescribePhase, PhaseNumber, ChooseStructure, SubmitCompass } from '../Create';
import { Form } from 'react-bootstrap';
configure({adapter:new Adapter()});

describe("DescribePhase",()=>{

    let mountedDescribePhase;
    let state;
    let props;

    const describePhase = () => {
        if (!mountedDescribePhase) mountedDescribePhase = mount(<DescribePhase {...props} />)
        return mountedDescribePhase;
    }

    beforeEach(() => {
        state = {
          title: "",
          description: ""
        };
        props = {
            createPhase: undefined,
            id: undefined,
            key: undefined
        };
        mountedDescribePhase = undefined;
    });

    describe("When DescribePhase is rendered correctly by snapshot",()=>{
        const DescribePhaseRendered = renderer.create(<DescribePhase />).toJSON();
        it("SnapShot matching",()=>{
            expect(DescribePhaseRendered).toMatchSnapshot();
        });
    
    })

    describe("When DescribePhase is rendered with empty state/props",()=>{
        beforeEach(() => {
            state = {
              title: "",
              description: ""
            };
            props = {
                createPhase: undefined,
                id: undefined,
                key: undefined
            };
        });
    
        it("Check DescribePhase Render with a blank state (UT)",()=>{
            const DescribePhaseComponent = describePhase()
            expect(DescribePhaseComponent.state()).toEqual(state);
        });

        it("Check DescribePhase Render with a blank prop (UT)",()=>{
            const DescribePhaseComponent = describePhase()
            expect(DescribePhaseComponent.props()).toEqual(props);
        });
    
    })

    describe("When props are being passed to DescribePhase",()=>{
        beforeEach(() => {
            state = {
                title: "",
                description: ""
            };
            props = {
                id: 0,
                createPhase: jest.fn(),
            };

        });

        it("renders a `Form` (UT)", () => {
            expect(describePhase().find(Form.Label).length).toBe(1);
        });
    
        it("Check DescribePhase Phase Label is correctly rendered with the right data (UT)",()=>{
            const PhaseLabel = describePhase().find(Form.Label)
            expect(PhaseLabel.props().children).toEqual("Phase 1");
        });

        it("Check DescribePhase is calling onChange correctly (IT)",()=>{
            const titleEvent = {
                target: { name:"title", value: 'this is a title' }
            };
            const descriptionEvent = {
                target: { name:"description", value: 'this is a description' }
            };

            const DescribePhase = describePhase();
            const spy = jest.spyOn(DescribePhase.instance(), "onChange");
            
            DescribePhase.instance().forceUpdate()
            DescribePhase.update();

            const TitleInput = DescribePhase.find("input[name='title']");
            const DescriptionInput = DescribePhase.find("textarea[name='description']");

            TitleInput.simulate('change', titleEvent);
            DescriptionInput.simulate('change', descriptionEvent);

            expect(spy.mock.calls.length).toBe(2);

            spy.mockRestore();
        });
    
    })

})

describe("PhaseNumber",()=>{

    let mountedPhaseNumber;
    let props;

    const phaseNumber = () => {
        if (!mountedPhaseNumber) mountedPhaseNumber = mount(<PhaseNumber {...props} />)
        return mountedPhaseNumber;
    }

    beforeEach(() => {
        props = {
            handlePhaseAmount: undefined,
        };
        mountedPhaseNumber = undefined;
    });

    describe("When PhaseNumber is rendered with empty props",()=>{
        beforeEach(() => {
            props = {
                handlePhaseAmount: undefined,
            };
        });

        it("Check PhaseNumber Render with a blank prop (UT)",()=>{
            const PhaseNumberComponent = phaseNumber()
            expect(PhaseNumberComponent.props()).toEqual(props);
        });
    })

    describe("When props are being passed to PhaseNumber",()=>{
        beforeEach(() => {
            props = {
                handlePhaseAmount: jest.fn( ),
            };
        });

        it("renders a `Form` (UT)", () => {
            expect(phaseNumber().find(Form.Control).length).toBe(1);
        });
    
        it("Check PhaseNumber Phase Label is correctly rendered with the right amount of options (UT)",()=>{
            const NumberForm = phaseNumber().find(Form.Control)
            expect(NumberForm.props().children.length).toEqual(12);
        });

        it("Check PhaseNumber is calling handlePhaseAmount when a selection has been made (IT)",()=>{
            const selectEvent = { target: { value: '2'} };

            const PhaseNumber = phaseNumber();

            const PhaseSelector = PhaseNumber.find("select");

            expect(props.handlePhaseAmount.mock.calls.length).toBe(0);

            PhaseSelector.simulate('change', selectEvent);

            expect(props.handlePhaseAmount.mock.calls.length).toBe(1);
        })
    })
})

describe("ChooseStructure",()=>{

    let mountedChooseStructure;
    let props;

    const chooseStructure = () => {
        if (!mountedChooseStructure) mountedChooseStructure = mount(<ChooseStructure {...props} />)
        return mountedChooseStructure;
    }

    beforeEach(() => {
        props = {
            handleCompassType: undefined,
        };
        mountedChooseStructure = undefined;
    });

    describe("When ChooseStructure is rendered with empty props",()=>{
        beforeEach(() => {
            props = {
                handleCompassType: undefined,
            };
        });

        it("Check ChooseStructure Renders with blank props (UT)",()=>{
            const ChooseStructureComponent = chooseStructure()
            expect(ChooseStructureComponent.props()).toEqual(props);
        });
    
    })

    describe("When props are being passed to ChooseStructure",()=>{
        beforeEach(() => {
            props = {
                handleCompassType: jest.fn(),
            };
        });

        it("renders two Buttons with the correct props (UT)", () => {
            expect(chooseStructure().find('button[id="Default"]').length).toBe(1);
            expect(chooseStructure().find('button[id="Custom"]').length).toBe(1);
        });

        it("Check chooseStructure's handleCompassType is being called (IT)",()=>{
            const ChooseStructure = chooseStructure();
            const CustomButton    = ChooseStructure.find('button[id="Custom"]')
            const DefaultButton   = ChooseStructure.find('button[id="Default"]')

            CustomButton.simulate('click');
            expect(props.handleCompassType).toBeCalledWith(false);

            DefaultButton.simulate('click');
            expect(props.handleCompassType).toBeCalledWith(true);

        });
    
    })
})

describe("SubmitCompass",()=>{

    let mountedSubmitCompass;
    let props;

    const submitCompass = () => {
        if (!mountedSubmitCompass) mountedSubmitCompass = mount(<SubmitCompass {...props} />)
        return mountedSubmitCompass;
    }

    beforeEach(() => {
        props = {
            createCompass: undefined,
            title: undefined,
            onChange: undefined
        };
        mountedSubmitCompass = undefined;
    });

    describe("When SubmitCompass is rendered with empty props",()=>{
        beforeEach(() => {
            props = {
                createCompass: undefined,
                title: undefined,
                onChange: undefined
            };
        });

        it("Check submitCompass Renders with blank props (UT)",()=>{
            const SubmitCompassComponent = submitCompass()
            expect(SubmitCompassComponent.props()).toEqual(props);
        });
    
    })

    describe("When props are being passed to SubmitCompass ",()=>{
        beforeEach(() => {
            props = {
                createCompass: jest.fn(),
                title: "",
                onChange: jest.fn()
            };
        });

        it("renders submit button and input (UT)", () => {
            const SubmitCompassComponent = submitCompass()
            expect(SubmitCompassComponent.find('input[name="title"]').length).toBe(1);
            expect(SubmitCompassComponent.find('button[className="input-button btn btn-primary"]').length).toBe(1);
        });

        it("Check SubmitCompassComponent's createCompass and onChange are being called (IT)",()=>{
            const SubmitCompassComponent = submitCompass()
            const TitleEvent = {target: {value: "Testing Title"}}
            const TitleInput   = SubmitCompassComponent.find('input[name="title"]')
            const SubmitButton = SubmitCompassComponent.find('button[className="input-button btn btn-primary"]')

            TitleInput.simulate('change',TitleEvent);
            expect(props.onChange).toBeCalled();
        
            SubmitButton.simulate('click');
            expect(props.createCompass).toBeCalled();

        });
    
    })

})

describe("CreatePage",()=>{

    let mountedCreatePage;
    let props;
    let state;

    const createPage = () => {
        if (!mountedCreatePage) mountedCreatePage = shallow(<CreatePage {...props} />)
        return mountedCreatePage;
    }

    beforeEach(() => {
        state = {
            compassTitle: "",
            canSubmit: false,
            defaultCompass: true,
            numberOfPhases: 0,
            phases: [],
            status: "chooseStructure",
            loading: false
          };
        props = {
            user: undefined,
        };
        mountedCreatePage = undefined;
    });

    describe("When CreatePage is rendered ",()=>{

        it("Check CreatePage Renders with a default state and loads Choose Structure first (UT)",()=>{
            const CreatePageComponent = createPage()

            expect(CreatePageComponent.state()).toEqual(state);
            expect(CreatePageComponent.find(ChooseStructure).length).toBe(1)
        });

        it("Check setting the state loads different phases through handleForms (IT)",()=>{
            const CreatePageComponent = createPage()
            const spy = jest.spyOn(CreatePageComponent.instance(), "handleForms");

            expect(CreatePageComponent.find(ChooseStructure).length).toBe(1)
            expect(spy.mock.calls.length).toBe(0)

        });
    })

    describe("When CreatePage changes forms ",()=>{

        it("Check setting the state loads different phases through handleForms (IT)",()=>{
            const CreatePageComponent = createPage()
            const spy = jest.spyOn(CreatePageComponent.instance(), "handleForms");

            CreatePageComponent.setState({status: "chooseStructure"})
            expect(CreatePageComponent.find(ChooseStructure).length).toBe(1)
            expect(spy.mock.calls.length).toBe(1)

            CreatePageComponent.setState({status: "numOfPhases",numberOfPhases:1})
            expect(CreatePageComponent.find(PhaseNumber).length).toBe(1)
            expect(spy.mock.calls.length).toBe(2)

            CreatePageComponent.setState({status: "createPhases"})
            expect(CreatePageComponent.find(DescribePhase).length).toBe(1)
            expect(spy.mock.calls.length).toBe(3)

            CreatePageComponent.setState({status: "submitCompass"})
            expect(CreatePageComponent.find(SubmitCompass).length).toBe(1)
            expect(spy.mock.calls.length).toBe(4)

        });
    })
})