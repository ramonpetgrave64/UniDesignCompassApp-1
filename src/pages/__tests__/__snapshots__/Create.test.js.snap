// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DescribePhase When DescribePhase is rendered correctly by snapshot SnapShot matching 1`] = `
<div
  className="form-group"
>
  <label
    className="form-sub-title form-label"
  >
    Phase NaN
  </label>
  <input
    className="form-control"
    id="phaseTitle"
    name="title"
    onChange={[Function]}
    placeholder="Title"
    type="text"
    value=""
  />
  <textarea
    className="form-control"
    id="phaseDescription"
    name="description"
    onChange={[Function]}
    placeholder="Description"
    rows="3"
    value=""
  />
</div>
`;
